### PR TITLE
[JENKINS-34466] Fix PCT against 2.0

### PR DESCRIPTION
**Pull request by andresrc**
When building against Jenkins 2.0 Groovy 2.4.6 is used instead of 1.8.9 in the baseline. This brings a couple of issues:
* For some methods in the whitelist there are additional overloads for more specific type arguments, generating approval errors.
* The behavior for ambiguous overloading has changed: no longer "some" version of the method is picked but an exception is thrown.

Besides, the PR includes:
* Findbugs fixes when building against 2.0
* Some cleanup after https://github.com/jenkinsci/script-security-plugin/pull/53

@reviewbybees

### CODE REVIEW

File: src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist [hunk 1]
```diff
@@ -127,7 +127,14 @@ staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods asBoolean java.lan
 staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods collect java.lang.Object groovy.lang.Closure
 staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods collect java.util.Collection groovy.lang.Closure
 staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods disjoint java.util.Collection java.util.Collection
+g2 staticMethod org.codehaus.groovy.runtime.DefaultGroovyMethods each java.lang.Iterable groovy.lang.Closure
```
Comment (andresrc) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: New overloads in Groovy 2.4.6 not in 1.8.9

File: src/test/java/org/jenkinsci/plugins/scriptsecurity/scripts/AbstractApprovalTest.java [hunk 2]
```diff
@@ -39,13 +39,9 @@
...
-        configureSecurity();
-        create().assertCannotUse();
-    }
-
-    @Test public void withSecurityAccepted() throws Exception {
+    @Test public void withSecurity() throws Exception {
         configureSecurity();
+        // Cannot use until approved
         create().assertCannotUse().approve().use();
     }
```
Comment (andresrc) [inline src/test/java/org/jenkinsci/plugins/scriptsecurity/scripts/AbstractApprovalTest.java]: Cleanup from https://github.com/jenkinsci/script-security-plugin/pull/53

**Comment by armfergom** [comment]
This LGTM as far as I understand. 🐝

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/StaticWhitelist.java [hunk 3]
```diff
@@ -44,12 +44,15 @@
...
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import groovy.lang.GroovySystem;
 import org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException;
 
 /**
  * Whitelist based on a static file.
  */
 public final class StaticWhitelist extends EnumeratingWhitelist {
+    private static final String G2_PREFIX = "g2";
+    private static final boolean GROOVY2 = GroovySystem.getVersion().startsWith("2.");
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/StaticWhitelist.java]: :ant: I think you want to use `VersionNumber` or similar to apply the same behavior (unless and until overridden) in Groovy 3.x.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/StaticWhitelist.java [hunk 4]
```diff
@@ -79,6 +81,30 @@ public StaticWhitelist(String... lines) throws IOException {
         this(asList(lines));
     }
 
+    /**
+     * Filters a line, returning the content that must be processed.
+     * @param line Line to filter.
+     * @return {@code null} if the like must be skipped or the content to process if not.
+     */
+    static String filter(String line) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/StaticWhitelist.java]: use `@CheckForNull`

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/StaticWhitelist.java [hunk 5]
```diff
@@ -79,6 +81,30 @@ public StaticWhitelist(String... lines) throws IOException {
...
+        if (line == null) {
+            return null;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/StaticWhitelist.java]: :ant: AFAICT both callers always pass non-null, so delete this and make parameter `@Nonnull`

File: src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist (hunk 1 above)
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: :bug: this is not maintainable, and there are/will be other libraries with different versions (including Jenkins core). I suppose the only reason not to just add the signature with no prefix is so that `GenericWhitelistTest.sanity` does not complain? Then just add a separate `generic-whitelist-groovy2` file and do not include it in `sanity` (or include it only if we can detect that the Groovy version in the classpath is sufficiently new).

File: src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java [hunk 6]
```diff
@@ -492,10 +495,28 @@ private Unsafe() {}
...
+    @Test public void ambiguousOverloads18() {
+        if (GroovySystem.getVersion().startsWith("1.")) {
+            // In 1.8.9 Groovy selects one of these. How, I do not know.
+            assertAmbiguity("null");
+        } else {
+            throw new AssumptionViolatedException("Not using Groovy 1.x");
+        }
+    }
+
+    @Test(expected = GroovyRuntimeException.class) public void ambiguousOverloads24() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java]: :ant: rather use `assertEvaluate` with a script catching `GroovyRuntimeException`

Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java]: (meaning inline `assertAmbiguity` back into `ambiguousOverloads18`)

File: src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist (hunk 1 above)
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist]: (treating as a bug not an ant because this is a format change and thus not easily reverted)

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/scripts/ScriptApproval.java [hunk 7]
```diff
@@ -756,14 +757,24 @@ private synchronized void removePendingScript(String hash) {
         return pendingSignatures;
     }
 
+    // Added to account for new findbugs annotations in ExtensionList#get (PCT scenario)
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/scripts/ScriptApproval.java]: Under PCT, compilation and FindBugs would still be done against the old baseline, so why is this here?

Comment (andresrc) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/scripts/ScriptApproval.java]: For these plugins, besides the current scenario, we are also doing full builds against both baseline and target.

File: src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/StaticWhitelist.java [hunk 8]
```diff
@@ -43,13 +43,17 @@
...
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import groovy.lang.GroovySystem;
+import hudson.util.VersionNumber;
 import org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException;
 
 /**
  * Whitelist based on a static file.
  */
 public final class StaticWhitelist extends EnumeratingWhitelist {
+    static final boolean GROOVY2 = new VersionNumber(GroovySystem.getVersion()).compareTo(new VersionNumber("2")) >= 0;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/StaticWhitelist.java]: :ant: inline into `GenericWhitelistTest`, its only remaining usage.

**Comment by jglick** [comment]
:bee:
